### ISSUE TITLE
Make Functions CLI cross platform

### ISSUE THREAD

**Issue by ahmelsayed** [Azure/azure-functions-cli#13]
Note: this work is under active development even though it's in the "backlog" milestone. See https://github.com/Azure/Azure-Functions/issues/98

Update 7/7 by @lindydonna: This work is underway, and we hope to have a preview in the next month. Please follow this issue to track the progress: https://github.com/Azure/Azure-Functions/issues/98

**Comment by Bomret** [comment]
I wonder why it is necessary to upvote this to be implemented instead being a top priority? Isn't Azure about embracing devs from all platforms?

**Comment by SteveALee** [comment]
Is it really only ```cmd``` on Windows right now? I want to use from Bash on Windows.

**Comment by lindydonna** [comment]
@Bomret It is indeed a priority, but the implementation is not trivial. We're working on it, and just wanted to allow customers to weigh in.

**Comment by fiveisprime** [comment]
There are some parts of the CLI that _should_ already work cross-platform
(`$ func functions create`, `$ func azure` actions). Right now the main script
wont even load on *nix platforms due to line endings and an immediate platform
test.

How about making the functions scaffolding and Azure management pieces available
cross-platform as a first step with some friendly errors for the parts that
absolutely won't work on anything but Windows?

**Comment by lindydonna** [comment]
@fiveisprime Definitely! We have plans to add Functions management operations to the Azure CLI 2.0.

**Comment by kevinbayes** [comment]
I upgraded today and I see that the cli is still not cross platform. A really pity would be great to get scaffolding and management cross-platform as @fiveisprime mentioned.

**Comment by lindydonna** [comment]
@kevinbayes The work is still in progress, as the entire local runtime needs to be ported to .NET core. The Azure CLI 2.0 work is much smaller and will come sooner.

**Comment by fiveisprime** [comment]
@kevinbayes for what it's worth you can generate the scaffolding cross platform
@now by running

```shell
$ npm i -g yo generator-azurefunctions
$ yo azurefunctions
```

👍

**Comment by kevinbayes** [comment]
@fiveisprime yeah that's what I use currently, but thanks for the tip.

**Comment by ams0** [comment]
FWIW, on my mac this works (i had to dos2unix main.js first):
```
sudo mono /usr/local/lib/node_modules/azure-functions-cli/bin/func.exe init
```

**Comment by jdneo** [comment]
Hi @lindydonna , when can we use Functions management operations in Azure CLI 2.0?

**Comment by lindydonna** [comment]
@jdneo It's under active development. This issue tracks the progress: https://github.com/Azure/azure-cli/issues/2352

**Comment by cicorias** [comment]
@fiveisprime doesn't work well. Not sure what's missing. But the generator fails. 

```
cicorias@arrakis:~/tmp/az/fn$ yo azurefunctions

     _-----_     
    |       |    ╭──────────────────────────╮
    |--(o)--|    │   Welcome to the Azure   │
   `---------´   │   Functions generator!   │
    ( _´U`_ )    ╰──────────────────────────╯
    /___A___\   /
     |  ~  |     
   __'.___.'__   
 ´   `  |° ´ Y ` 

? Select an option... List templates by type
There are 8 event types available
? Select an event type... blobTrigger
There are 0 templates available
? Select from one of the available templates... 


readline.js:990
            throw err;
            ^

TypeError: Cannot read property 'value' of undefined
    at Prompt.getCurrentValue (/Users/cicorias/.nvm/versions/node/v7.10.0/lib/node_modules/yo/node_modules/yeoman-environment/node_modules/inquirer/lib/prompts/list.js:129:51)
    at tryCatcher (/Users/cicorias/.nvm/versions/node/v7.10.0/lib/node_modules/yo/node_modules/rx/dist/rx.js:63:31)
    at InnerObserver.next (/Users/cicorias/.nvm/versions/node/v7.10.0/lib/node_modules/yo/node_modules/rx/dist/rx.js:5407:43)
    at InnerObserver.Rx.internals.AbstractObserver.AbstractObserver.onNext (/Users/cicorias/.nvm/versions/node/v7.10.0/lib/node_modules/yo/node_modules/rx/dist/rx.js:1762:31)
    at InnerObserver.tryCatcher (/Users/cicorias/.nvm/versions/node/v7.10.0/lib/node_modules/yo/node_modules/rx/dist/rx.js:63:31)
    at AutoDetachObserverPrototype.next (/Users/cicorias/.nvm/versions/node/v7.10.0/lib/node_modules/yo/node_modules/rx/dist/rx.js:5883:51)
    at AutoDetachObserver.Rx.internals.AbstractObserver.AbstractObserver.onNext (/Users/cicorias/.nvm/versions/node/v7.10.0/lib/node_modules/yo/node_modules/rx/dist/rx.js:1762:31)
    at TakeObserver.next (/Users/cicorias/.nvm/versions/node/v7.10.0/lib/node_modules/yo/node_modules/rx/dist/rx.js:5646:17)
    at TakeObserver.Rx.internals.AbstractObserver.AbstractObserver.onNext (/Users/cicorias/.nvm/versions/node/v7.10.0/lib/node_modules/yo/node_modules/rx/dist/rx.js:1762:31)
    at TakeObserver.tryCatcher (/Users/cicorias/.nvm/versions/node/v7.10.0/lib/node_modules/yo/node_modules/rx/dist/rx.js:63:31)
    at AutoDetachObserverPrototype.next (/Users/cicorias/.nvm/versions/node/v7.10.0/lib/node_modules/yo/node_modules/rx/dist/rx.js:5883:51)
    at AutoDetachObserver.Rx.internals.AbstractObserver.AbstractObserver.onNext (/Users/cicorias/.nvm/versions/node/v7.10.0/lib/node_modules/yo/node_modules/rx/dist/rx.js:1762:31)
    at Subject.onNext (/Users/cicorias/.nvm/versions/node/v7.10.0/lib/node_modules/yo/node_modules/rx/dist/rx.js:5998:19)
    at Subject.tryCatcher (/Users/cicorias/.nvm/versions/node/v7.10.0/lib/node_modules/yo/node_modules/rx/dist/rx.js:63:31)
    at AutoDetachObserverPrototype.next (/Users/cicorias/.nvm/versions/node/v7.10.0/lib/node_modules/yo/node_modules/rx/dist/rx.js:5883:51)
    at AutoDetachObserver.Rx.internals.AbstractObserver.AbstractObserver.onNext (/Users/cicorias/.nvm/versions/node/v7.10.0/lib/node_modules/yo/node_modules/rx/dist/rx.js:1762:31)
cicorias@arrakis:~/tmp/az/fn$ 

```

**Comment by lindydonna** [comment]
We aren't maintain the Yeoman generator anymore, we have logic for this in the CLI itself

**Comment by penkin** [comment]
I'm really surprised that this is still not supported on other platforms. I'm on OS X.

**Comment by wind-rider** [comment]
I'm interested in hosting Functions on Raspberry Pi / Windows IoT Core.

**Comment by colegleason** [comment]
I am also looking for macOS support

**Comment by atrauzzi** [comment]
I'd just love to be able to set up a local development environment for them.

**Comment by sureshakella** [comment]
I would love to be able to set up a local development for Ubuntu > 16.04

**Comment by kinggeorge** [comment]
I just tweeted ( https://twitter.com/GeorgBuske/status/883141097790296064 ) about my positive first impression of Azure as Google Cloud Functions are way too beta in functionality - and now I come across this (very first link after google search for Linux CLI deployment), come on #MSFT - do I really need to stay with #AMZN or are there real alternatives (without windows build environment)? ;-(

**Comment by SteveALee** [comment]
@kinggeorge if you write your functions in one of the portable languages you can deploy to azure using FTP or fit intergration. So can have a non windows build CI/CD

What you don't yet have is access to the full CLI and local dev

**Comment by lindydonna** [comment]
@kinggeorge Exactly, you actually get the same experience as AWS when you're on a Mac/Linux. Lambda does not provide a local functions runtime at all.

The port to .NET Core is underway and we should have a preview within the next month.

**Comment by kinggeorge** [comment]
@SteveALee Thanks, I'll check that out later. 
@lindydonna Thanks for feedback. I admit it wasn't after deep analysis but very first impressions, first positive (as tweeted), then negative, I also wasn't referring to the emulator function (which non-the-less I like at Google Cloud Functions during tests), but to have a handy CLI tool for managing lambdas / functions (which AWS has). So it was just a bit frustrating to see the error message (`npm ERR! notsup Valid OS:    win32`...) after I was happy to see all the available features in the azure function dashboard, which Google was lacking or are a pain to setup [hack] (e.g. custom domain, CORS). Looking forward to the release of the rewrite... Cheers :)

**Comment by SteveALee** [comment]
I agree the deplyment options are minimal and not an ideal fit for a CO pipeline.

The current git support really requires an extra repository so deployment can be triggered by a check-in at the end of the pipeline. Hence the thought of using ftp whuch has push semantics.

You could fun the cli in a container ss part of the pipeline but windows containers are rarely supported which brings us back to need a linux version of the cli.

I guess we could create function that implement a Ci pipeline  (or logic apps) with azure container services but I don't think all the required hooks are there yet

**Comment by blown302** [comment]
+1

**Comment by atrauzzi** [comment]
@SteveALee Oh please no. Anything that requires more Azure services to be used as crutches is not a valid option.  Need to get out of these kludge loops. There is something concerning with the way Microsoft is approaching the "developer joy" story with Azure. I see so many shortcuts end up being applied that resort to more lock in.  Builds and deployments should be runnable from any platform in our outside of Azure using standard scripting options.

Also, no to Windows containers. That just misses the whole issue right now which is that the optimal deploy flow is only available for Windows (msdeploy). Break down those old MS habits and start thinking cross platform!

**Comment by cicorias** [comment]
@lindydonna could the Yo generator be flagged as deprecated and possibly updated one last time indicating deprecation? People find the old stuff and become frustrated and bail before they spelunk further and find updated bits. A real turnoff.

At first the readme. If you're ok I can send a PR

**Comment by SteveALee** [comment]
@atrauzzi I'd totally prefer to use stand tools like rsync but was just brainstorming work arounds :) Also teh Windows build flow is not optimal either.

We want a simple folder sync and FTP does NOT do that, at least not the common clients.

I investigated this and [blogged it in some detail](http://azureserverless.com/2017/08/12/deploying-to-azure-functions-from-npm-scripts/).

**Comment by lindydonna** [comment]
@cicorias Good catch! Yes, please do and we can update the package.

**Comment by carcarbe** [comment]
+1 We really need this feature.

**Comment by kwe** [comment]
+1 really import for us also

**Comment by ahmelsayed** [comment]
This has been released today see https://github.com/Azure/Azure-Functions/issues/98#issuecomment-331878210 for details 

> Today! 
> Announcement blogs: 
> - https://azure.microsoft.com/en-us/blog/serverless-for-all-developers-bringing-azure-functions-to-linux-mac-planet-scale-nosql-real-time-analytics-and-productivity-apps/
> - https://blogs.msdn.microsoft.com/appserviceteam/2017/09/25/develop-azure-functions-on-any-platform/

**Comment by ndarilek** [comment]
If this is indeed cross-platform, you may want to remove the "only works on Windows" bit from the README. :) I'm a Linux user and initially went away in search of another solution, then thought I'd come back here to vote up the issue as requested, and now I see that it should work.

**Comment by lindydonna** [comment]
CC @jeffholan

**Comment by lindydonna** [comment]
CC @jeffhollan to update the readme.

**Comment by joshuataylor** [comment]
Yeah, this doesn't work on Linux..

```
npm ERR! notsup Unsupported platform for azure-functions-core-tools@1.0.8: wanted {"os":"win32","arch":"any"} (current: {"os":"linux","arch":"x64"})
npm ERR! notsup Valid OS:    win32
npm ERR! notsup Valid Arch:  any
npm ERR! notsup Actual OS:   linux
npm ERR! notsup Actual Arch: x64
```

**Comment by EisenbergEffect** [comment]
The npm package still says this only supports Windows. What's the deal with this? Is it Windows-only or not?

**Comment by ahmelsayed** [comment]
I think the README for v1.0 got published for v2.0 tag. I'm not sure how, I'll fix the readme on the next release. The [package.json metadata](https://github.com/Azure/azure-functions-core-tools/blob/dev/src/Azure.Functions.Cli/npm/package.json#L20-L24) should list windows, linux and mac

**Comment by EisenbergEffect** [comment]
Thanks for the clarification @ahmelsayed !
